### ISSUE TITLE
[IMPROVEMENTS] Add faceIndex for XbimFaceTriangulation and IFCFACETEDBREP

### ISSUE THREAD

**Issue by liszto** [xBimTeam/XbimEssentials#202]
Hello guys,
I'm still working on all our XBim / ifc stuff and I'm facing an annoying issue when I try to create a nice rendering in our engine.
When I generate some asset that contains some "IFCFACETEDBREP", in XBimFaceTriangulation, each triangle generated based on an IFCFACETEDBREP is an independant face for you, so each triangle have its own normals.
So when the normals at one point are not the same for the same point it create some crappy display with rendering pipeline.
Is there a way to add in the XbimTriangulation a new parameter to allow us to link each triangle between them, like a faceIndex ?
In this way, it's possible to detect and retrieve all the triangle that could be merged into a single mesh.

Or maybe iterate through the faces in a shape would be enough if the IsPlanar value is different from 1 for each face ?

Thanks.

**Comment by martin1cerny** [comment]
If you can use IFC4 than you can just use [IfcTriangulatedFaceSet](http://www.buildingsmart-tech.org/ifc/IFC4/Add2/html/link/ifctriangulatedfaceset.htm). Normals in there will create smooth visual appearance.

**Comment by liszto** [comment]
mmm that's not the point.
Currently I'm trying to read a file which is not an ifc4 file format scheme.

And I'm trying to determine when a face is "flat" or not. For a moment I would use your IsPlanar method but it's not really what I need.

Because I found some counterexample like a wall with extrude like your sample house model.
The vertical wall that contains window is detected as a "non planar" element just because he just hasn't one normal for its face.

So basically,
even if all normal defines in the IList<XbimPackedNormal> are the same you will consider it as a non planar whereas it is in fact.

**Comment by martin1cerny** [comment]
Well, it is virtually impossible to say if the surface should be smoothed out or not if it is not in the original data. You are right that we should consider face to be planar if all vertices are in the same plane and all normals point in the direction parallel with the plane normal. 

But if you just have a pile of triangles which come from IfcFacetedBRep you can't assume anything. Smoothing applied to all faces would create very bad visual impression as square would look like a sphere and for example wall corners would become invisible.

**Comment by liszto** [comment]
I'm totally agree with your last point. 
Currently, all my IfcFacetedBRep are curved so it was logical according to me to smooth them by applying vertex merging and normal merging in this wayy all would be rendered as a smooth face.
But I think you have some exemple where people will generate an IfcFacetedBRep with some right angle or something close from it AND in this particular case you won't smooth it.

So for me there is maybe multiple thing to do : 
1. Redefine what is really a planar face by using the vertices on the same plane and normal orientation per vertex pricks up in the same direction.

2. I think (in my case) it would be useful to smoothby merging vertex / normal at a same point when the angle between 2 normals in lower than a particular value. In this way I can try to manage those particular case between smooth vs angular rendering.

**Comment by martin1cerny** [comment]
Good example will be any wall cut by a curved roof exported from Revit. That will come out as `IfcFacetedBRep` and will have any arbitrary angle on top. But should be represented without normal smoothing.

**Comment by martin1cerny** [comment]
Well, depends on the version. [Here](https://github.com/xBimTeam/XbimEssentials/files/2339768/SampleHouse.zip) is a one where at least one of the external walls (3cUkl32yn9qRSPvBJVyWw5) is `IfcFacetedBRep`.

**Comment by liszto** [comment]
Thanks @martin1cerny !
I'll try to check this this afternoon cause it's my last day before my 2 holiday weeks.

**Comment by liszto** [comment]
Ok I test it midday and the result seems good too !
![normal_rendering](https://user-images.githubusercontent.com/408031/44912211-bcf16080-ad29-11e8-9f92-baa0519de9a2.gif)
This one is done with your file just above

**Comment by martin1cerny** [comment]
I assume you didn't apply the smoothing of normals here. Did you? Or do you only smooth normals within one face?

For the broken normals I didn't see this in our tools. So, may it be something specific to your implementation?

**Comment by andyward** [comment]
What was the outcome on this. Can we close?

**Comment by liszto** [comment]
yeah I think you can. I currently on dissociated tasks so I put all my ifc task on hold til the end of the year I think.
If I need to reopen it, I will, so I close ;=)
